File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/ConcatenateEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;

/*
+(left String, right String) String

The concatenate (+) operator performs string concatenation of its arguments.
If either argument is null, the result is null.
*/

public class ConcatenateEvaluator {

    public static Object concatenate(Object left, Object right) {
        if (left == null || right == null) {
            return null;
        }

        if(left instanceof String && right instanceof String){
            return ((String)left).concat((String)right);
        }

        throw new InvalidOperatorArgument(
                "Concatenate(String, String)",
                String.format("Concatenate(%s, %s)", left.getClass().getName(), right.getClass().getName())
        );
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/elm/executing/ConcatenateEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument

/*
+(left String, right String) String

The concatenate (+) operator performs string concatenation of its arguments.
If either argument is null, the result is null.
*/
object ConcatenateEvaluator {
    @JvmStatic
    fun concatenate(left: Any?, right: Any?): Any? {
        if (left == null || right == null) {
            return null
        }

        if (left is String && right is String) {
            return left + right
        }

        throw InvalidOperatorArgument(
            "Concatenate(String, String)",
            String.format("Concatenate(%s, %s)", left.javaClass.name, right.javaClass.name),
        )
    }
}
